pkg/autoscaling/calculation.go

```diff
@@ -0,0 +1,210 @@
+		return nil
+	}
+	return plans
+}
+
+func getPlans(rc *cluster.RaftCluster, strategy *Strategy, component ComponentType) ([]*Plan, uint64) {
+	var instances []string
+	if component == TiKV {
+		instances = filterTiKVInstances(rc)
+	} else {
+		instances = getTiDBInstances()
+	}
+
+	if len(instances) == 0 {
+		return nil, 0
+	}
+
+	totalCPUUseTime := getTotalCPUUseTime(component, instances, MetricsTimeDuration)
+	currentQuota := getTotalCPUQuota(component, instances)
+	totalCPUTime := float64(currentQuota) * MetricsTimeDuration.Seconds()
+	usage := totalCPUUseTime / totalCPUTime
+	maxThreshold, minThreshold := getCPUThresholdByComponent(strategy, component)
+
+	if usage > maxThreshold {
+		scaleOutQuota := totalCPUUseTime/maxThreshold - totalCPUTime
+		return calculateScaleOutPlan(rc, strategy, component, scaleOutQuota, currentQuota)
+	}
+	if usage < minThreshold {
+		scaleInQuota := totalCPUUseTime/minThreshold - totalCPUTime
+		return calculateScaleInPlan(rc, strategy, component, scaleInQuota), 0
+	}
+	return nil, 0
+}
+
+func filterTiKVInstances(rc *cluster.RaftCluster) []string {
+	var instances []string
+	stores := rc.GetStores()
+	for _, store := range stores {
+		if store.GetState() == metapb.StoreState_Up {
+			instances = append(instances, store.GetAddress())
+		}
+	}
+	return instances
+}
+
+// TODO: get TiDB instances
+func getTiDBInstances() []string {
+	return []string{}
+}
+
+// TODO: get total CPU use time through Prometheus.
+func getTotalCPUUseTime(component ComponentType, instances []string, duration time.Duration) float64 {
+	return 1.0
+}
+
+// TODO: get total CPU quota through Prometheus.
+func getTotalCPUQuota(component ComponentType, instances []string) uint64 {
+	return 1
+}
+
+func getCPUThresholdByComponent(strategy *Strategy, component ComponentType) (maxThreshold float64, minThreshold float64) {
+	for _, rule := range strategy.Rules {
+		if rule.Component == component.String() {
+			return rule.CPURule.MaxThreshold, rule.CPURule.MinThreshold
+		}
+	}
+	return 0, 0
+}
+
+func getResourcesByComponent(strategy *Strategy, component ComponentType) []*Resource {
+	var resTyp []string
+	var resources []*Resource
+	for _, rule := range strategy.Rules {
+		if rule.Component == component.String() {
+			resTyp = rule.CPURule.ResourceTypes
+		}
+	}
+	for _, res := range strategy.Resources {
+		for _, typ := range resTyp {
+			if res.ResourceType == typ {
+				resources = append(resources, res)
+			}
+		}
+	}
+	return resources
+}
+
+// TODO: implement calculateScaleOutPlan details
+func calculateScaleOutPlan(rc *cluster.RaftCluster, strategy *Strategy, component ComponentType, scaleOutQuota float64, currentQuota uint64) ([]*Plan, uint64) {
+	groups := getScaledGroupsByComponent(component)
+	group := findBestGroupToScaleOut(rc, strategy, scaleOutQuota, groups, component)
+
+	resCPU := float64(getCPUByResourceType(strategy, group.ResourceType))
+	scaleOutCount := uint64(math.Ceil(scaleOutQuota / resCPU))
```

Comment:
I think currently we can simplify the logic as if the auto-scaling instances count changing is always 1 which means If the usage rate > max threshold, we will add one instance.

Reply:
Is it better to make step configurable?

Reply:
Making step configurable is ok to me but it will make the logic complex which against our initial goal. We can support it later.

Reply:
Making step configurable is ok to me too

Reply:
Making step configurable is ok to me too